A database (?) error on pyemma.coordinates.source()

Hit a problem while trying to do k-means on my TICA-projected trajs, `pyemma.coordinates.source(glob.glob('*.npy'))` errors, while it worked perfectly ok on the featurized trajs (arrays are even same dimensions, keeping all TICs). 

Here's the error:
```python
In [108]: source = pyemma.coordinates.source(glob.glob('*.npy'))
Obtaining file info:   1% (  76/5020) [#                                                                             ] eta 03:02 |13-06-17 12:58:31 pyemma.coordinates.data.util.traj_info_cache INFO     Cleaning database because it has too much entries or is too large.
Entries: 50084. Size: 24.15MB. Configured max_entires: 50000. Max_size: 500MB
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/cbio/jclab/home/rafal.wiewiora/anaconda3/lib/python3.6/site-packages/pyemma/coordinates/data/util/traj_info_cache.py in __getitem__(self, filename_reader_tuple)
    186         try:
--> 187             info = self._database.get(key)
    188             if not isinstance(info, TrajInfo):

/cbio/jclab/home/rafal.wiewiora/anaconda3/lib/python3.6/site-packages/pyemma/coordinates/data/util/traj_info_backends.py in get(self, key)
    222         if not row:
--> 223             raise KeyError()
    224         info = self._create_traj_info(row)

KeyError: 

During handling of the above exception, another exception occurred:

AssertionError                            Traceback (most recent call last)
<ipython-input-108-9d3c8ffd8f2f> in <module>()
----> 1 source = pyemma.coordinates.source(glob.glob('*.npy'))

/cbio/jclab/home/rafal.wiewiora/anaconda3/lib/python3.6/site-packages/pyemma/coordinates/api.py in source(inp, features, top, chunk_size, **kw)
    346             isinstance(inp, (list, tuple))
    347             and (any(isinstance(item, (list, tuple, _string_types)) for item in inp) or len(inp) is 0)):
--> 348         reader = create_file_reader(inp, top, features, chunk_size=chunk_size if chunk_size is not None else 100, **kw)
    349 
    350     elif isinstance(inp, _np.ndarray) or (isinstance(inp, (list, tuple))

/cbio/jclab/home/rafal.wiewiora/anaconda3/lib/python3.6/site-packages/pyemma/coordinates/data/util/reader_utils.py in create_file_reader(input_files, topology, featurizer, chunk_size, **kw)
    103                 else:
    104                     if suffix in ['.npy', '.npz']:
--> 105                         reader = NumPyFileReader(input_list, chunksize=chunk_size)
    106                     # otherwise we assume that given files are ascii tabulated data
    107                     else:

/cbio/jclab/home/rafal.wiewiora/anaconda3/lib/python3.6/site-packages/pyemma/coordinates/data/numpy_filereader.py in __init__(self, filenames, chunksize, mmap_mode)
     62 
     63         self.mmap_mode = mmap_mode
---> 64         self.filenames = filenames
     65 
     66     def _create_iterator(self, skip=0, chunk=0, stride=1, return_trajindex=False, cols=None):

/cbio/jclab/home/rafal.wiewiora/anaconda3/lib/python3.6/site-packages/pyemma/coordinates/data/_base/datasource.py in filenames(self, filename_list)
    114             for filename in filename_list:
    115                 if config.use_trajectory_lengths_cache:
--> 116                     info = TrajectoryInfoCache.instance()[filename, self]
    117                 else:
    118                     info = self._get_traj_info(filename)

/cbio/jclab/home/rafal.wiewiora/anaconda3/lib/python3.6/site-packages/pyemma/coordinates/data/util/traj_info_cache.py in __getitem__(self, filename_reader_tuple)
    204             info.abs_path = abs_path
    205             # store info in db
--> 206             self.__setitem__(info)
    207 
    208             # save forcefully now

/cbio/jclab/home/rafal.wiewiora/anaconda3/lib/python3.6/site-packages/pyemma/coordinates/data/util/traj_info_cache.py in __setitem__(self, traj_info)
    245 
    246     def __setitem__(self, traj_info):
--> 247         self._database.set(traj_info)
    248 
    249     def clear(self):

/cbio/jclab/home/rafal.wiewiora/anaconda3/lib/python3.6/site-packages/pyemma/coordinates/data/util/traj_info_backends.py in set(self, traj_info)
    215                             % (self.num_entries, (current_size*1.0 / 1024**2),
    216                                config.traj_info_max_entries, config.traj_info_max_size))
--> 217                 self._clean(n=self.clean_n_entries)
    218 
    219     def get(self, key):

/cbio/jclab/home/rafal.wiewiora/anaconda3/lib/python3.6/site-packages/pyemma/coordinates/data/util/traj_info_backends.py in _clean(self, n)
    349         cur = self._database.execute(stmnt)
    350         self._database.commit()
--> 351         assert cur.rowcount == len(ids), "deleted not as many rows(%s) as desired(%s)" %(cur.rowcount, len(ids))
    352 
    353         # iterate over all LRU databases and delete those ids, we've just deleted from the main db.

AssertionError: deleted not as many rows(1) as desired(15025)
```

Interestingly, this hack makes it work:

```python
fnames = glob.glob('*.npy')

for i in np.arange(1, len(fnames)):                   
    try:
        source = pyemma.coordinates.source(fnames[0:i])
    except:
        source = pyemma.coordinates.source(fnames[0:i])
```

(i.e keep making the list of files longer by one and repeat when it fails) - I tried this once, started failing at 43 files, and then on every single file addition if fails once, works another time and so on:

```python
Obtaining file info:  97% (42/43) [################################################################################  ] eta 00:01 |13-06-17 12:56:33 pyemma.coordinates.data.util.traj_info_cache INFO     Cleaning database because it has too much entries or is too large.
Entries: 50057. Size: 24.15MB. Configured max_entires: 50000. Max_size: 500MB
Obtaining file info:  97% (43/44) [################################################################################  ] eta 00:01 -13-06-17 12:56:35 pyemma.coordinates.data.util.traj_info_cache INFO     Cleaning database because it has too much entries or is too large.
(...)
```

Only got it up to 88 files now, and keeps doing this fail-work pattern. I think I'm going to try running with this hack on all 5020 trajectories and seeing if that works (EDIT: this of course turned out to take way too much time to be reasonable) and if I can get those k-means running asap.

--- a/doc/source/CHANGELOG.rst
+++ b/doc/source/CHANGELOG.rst
@@ -18,7 +18,7 @@ Changelog
 
 - datasets: fixed get_multi_temperature_data and get_umbrella_sampling_data for Python 3. #1102
 - coordinates: fixed StreamingTransformers (TICA, Kmeans, etc.) not respecting the in_memory flag. #1112
-
+- coordinates: made TrajectoryInfoCache more fail-safe in case of concurrent processes. #1122
 
 2.4 (05-19-2017)
 ----------------

--- a/pyemma/coordinates/data/util/traj_info_backends.py
+++ b/pyemma/coordinates/data/util/traj_info_backends.py
@@ -283,9 +283,10 @@ class SqliteDB(AbstractDB):
                 # if there are many jobs to write to same database at same time, the timeout could be hit
                 logger.debug('could not update LRU info for db %s', db_name)
 
-        import threading
-        threading.Thread(target=_update).start()
-
+        # this could lead to another (rare) race condition during cleaning...
+        #import threading
+        #threading.Thread(target=_update).start()
+        _update()
 
     @staticmethod
     def _create_traj_info(row):

--- a/pyemma/coordinates/data/util/traj_info_backends.py
+++ b/pyemma/coordinates/data/util/traj_info_backends.py
@@ -358,17 +358,17 @@ class SqliteDB(AbstractDB):
 
         sql_compatible_ids = SqliteDB._format_tuple_for_sql(ids)
 
-        stmnt = "DELETE FROM traj_info WHERE hash in (%s)" % sql_compatible_ids
-        cur = self._database.execute(stmnt)
-        self._database.commit()
-        assert cur.rowcount == len(ids), "deleted not as many rows(%s) as desired(%s)" %(cur.rowcount, len(ids))
+        with self._database as c:
+            c.execute("DELETE FROM traj_info WHERE hash in (%s)" % sql_compatible_ids)
 
-        # iterate over all LRU databases and delete those ids, we've just deleted from the main db.
-        age_by_hash.sort(key=itemgetter(2))
-        for db, values in itertools.groupby(age_by_hash, key=itemgetter(2)):
-            values = tuple(v[0] for v in values)
-            with sqlite3.connect(db, timeout=self.lru_timeout) as conn:
-                    stmnt = "DELETE FROM usage WHERE hash IN (%s)" \
-                            % SqliteDB._format_tuple_for_sql(values)
-                    curr = conn.execute(stmnt)
-                    assert curr.rowcount == len(values), curr.rowcount
+            # iterate over all LRU databases and delete those ids, we've just deleted from the main db.
+            # Do this within the same execution block of the main database, because we do not want the entry to be deleted,
+            # in case of a subsequent failure.
+            age_by_hash.sort(key=itemgetter(2))
+            for db, values in itertools.groupby(age_by_hash, key=itemgetter(2)):
+                values = tuple(v[0] for v in values)
+                with sqlite3.connect(db, timeout=self.lru_timeout) as conn:
+                        stmnt = "DELETE FROM usage WHERE hash IN (%s)" \
+                                % SqliteDB._format_tuple_for_sql(values)
+                        curr = conn.execute(stmnt)
+                        assert curr.rowcount == len(values), curr.rowcount

--- a/pyemma/coordinates/data/util/traj_info_backends.py
+++ b/pyemma/coordinates/data/util/traj_info_backends.py
@@ -119,9 +119,11 @@ class SqliteDB(AbstractDB):
         # Converts TEXT to np.array when selecting
         sqlite3.register_converter("NPARRAY", convert_array)
         self._database = sqlite3.connect(filename if filename is not None else ":memory:",
-                                         detect_types=sqlite3.PARSE_DECLTYPES, timeout=1000*1000,
+                                         detect_types=sqlite3.PARSE_DECLTYPES, timeout=5,
                                          isolation_level=None)
         self.filename = filename
+
+        self.lru_timeout = 5.0 # python sqlite3 specifies timeout in seconds instead of milliseconds.
 
         try:
             cursor = self._database.execute("select num from version")
@@ -148,8 +150,8 @@ class SqliteDB(AbstractDB):
 
     def _create_new_db(self):
         # assumes self.database is a sqlite3.Connection
-        create_version_table = "CREATE TABLE version (num INTEGER PRIMARY KEY);"
-        create_info_table = """CREATE TABLE traj_info(
+        create_version_table = "CREATE TABLE IF NOT EXISTS version (num INTEGER PRIMARY KEY);"
+        create_info_table = """CREATE TABLE IF NOT EXISTS  traj_info(
             hash VARCHAR(64) PRIMARY KEY,
             length INTEGER,
             ndim INTEGER,
@@ -176,8 +178,13 @@ class SqliteDB(AbstractDB):
 
     @db_version.setter
     def db_version(self, val):
-        self._database.execute("insert into version VALUES (?)", [val])
-        self._database.commit()
+        import sqlite3
+        with self._database:
+            try:
+                self._database.execute("insert into version VALUES (?)", [val])
+            except sqlite3.IntegrityError:
+                pass
+       # self._database.commit()
 
     @property
     def num_entries(self):
@@ -197,11 +204,11 @@ class SqliteDB(AbstractDB):
         statement = ("INSERT INTO traj_info (hash, length, ndim, offsets, abs_path, version, lru_db)"
                      "VALUES (?, ?, ?, ?, ?, ?, ?)", values)
         try:
-             self._database.execute(*statement)
+            with self._database as c:
+                c.execute(*statement)
         except sqlite3.IntegrityError as ie:
             logger.exception("insert failed: %s " % ie)
             return
-        self._database.commit()
 
         self._update_time_stamp(hash_value=traj_info.hash_value)
 
@@ -257,20 +264,28 @@ class SqliteDB(AbstractDB):
         if not db_name:
             db_name=':memory:'
 
-        import sqlite3
+        def _update():
+            import sqlite3
+            try:
+                with sqlite3.connect(db_name, timeout=self.lru_timeout) as conn:
+                    """ last_read is a result of time.time()"""
+                    conn.execute('CREATE TABLE IF NOT EXISTS usage '
+                                 '(hash VARCHAR(32), last_read FLOAT)')
+                    conn.commit()
+                    cur = conn.execute('select * from usage where hash=?', (hash_value,))
+                    row = cur.fetchone()
+                    if not row:
+                        conn.execute("insert into usage(hash, last_read) values(?, ?)", (hash_value, time.time()))
+                    else:
+                        conn.execute("update usage set last_read=? where hash=?", (time.time(), hash_value))
+                    conn.commit()
+            except sqlite3.OperationalError:
+                # if there are many jobs to write to same database at same time, the timeout could be hit
+                logger.debug('could not update LRU info for db %s', db_name)
 
-        with sqlite3.connect(db_name) as conn:
-            """ last_read is a result of time.time()"""
-            conn.execute('CREATE TABLE IF NOT EXISTS usage '
-                         '(hash VARCHAR(32), last_read FLOAT)')
-            conn.commit()
-            cur = conn.execute('select * from usage where hash=?', (hash_value,))
-            row = cur.fetchone()
-            if not row:
-                conn.execute("insert into usage(hash, last_read) values(?, ?)", (hash_value, time.time()))
-            else:
-                conn.execute("update usage set last_read=? where hash=?", (time.time(), hash_value))
-            conn.commit()
+        import threading
+        threading.Thread(target=_update).start()
+
 
     @staticmethod
     def _create_traj_info(row):
@@ -324,10 +339,8 @@ class SqliteDB(AbstractDB):
 
         # debug: distribution
         len_by_db = {os.path.basename(db): len(hashs_by_db[db]) for db in hashs_by_db.keys()}
-        logger.debug("distribution of lru: %s" % str(len_by_db))
+        logger.debug("distribution of lru: %s", str(len_by_db))
         ### end dbg
-
-        self.lru_timeout = 1000 #1 sec
 
         # collect timestamps from databases
         for db in hashs_by_db.keys():

--- a/pyemma/coordinates/tests/test_traj_info_cache.py
+++ b/pyemma/coordinates/tests/test_traj_info_cache.py
@@ -57,7 +57,7 @@ class TestTrajectoryInfoCache(unittest.TestCase):
         config.use_trajectory_lengths_cache = True
 
     def setUp(self):
-        self.work_dir = tempfile.mkdtemp("traj_cache_test")
+        self.work_dir = tempfile.mkdtemp(prefix="traj_cache_test")
         self.tmpfile = tempfile.mktemp(dir=self.work_dir)
         self.db = TrajectoryInfoCache(self.tmpfile)
 
@@ -337,6 +337,37 @@ class TestTrajectoryInfoCache(unittest.TestCase):
                 config.cfg_dir = old_cfg_dir
             except ConfigDirectoryException:
                 pass
+
+    def test_stress(self):
+        arrays = [np.empty((5, 2))] * 100
+        npy_files = [os.path.join(self.work_dir, '{}.npy'.format(i)) for i in range(len(arrays))]
+        [np.save(f, x) for f, x in zip(npy_files, arrays)]
+        env = os.environ.copy()
+        env['PYEMMA_CFG_DIR'] = self.work_dir
+        import subprocess
+        import sys
+        import time
+        script = 'import pyemma; assert pyemma.config.cfg_dir=="{cfg_dir}", pyemma.config.cfg_dir; pyemma.coordinates.source({files})' \
+            .format(cfg_dir=self.work_dir, files=npy_files)
+        failed = False
+        procs = [subprocess.Popen([sys.executable, '-c', script], env=env) for _ in range(10)]
+        while procs:
+            for proc in procs:
+                retcode = proc.poll()
+                if retcode != None:
+                    procs.remove(proc)
+                    break
+                else:  # No process is done, wait a bit and check again.
+                    time.sleep(.1)
+                    continue
+
+            # Here, `proc` has finished with return code `retcode`
+            if retcode is not None and retcode != 0:
+                print('process failed with {}'.format(retcode))
+                failed = True
+                break
+
+        self.assertTrue(not failed)
 
 if __name__ == "__main__":
     unittest.main()

--- a/pyemma/coordinates/tests/test_traj_info_cache.py
+++ b/pyemma/coordinates/tests/test_traj_info_cache.py
@@ -351,10 +351,12 @@ class TestTrajectoryInfoCache(unittest.TestCase):
             .format(cfg_dir=self.work_dir, files=npy_files)
         failed = False
         procs = [subprocess.Popen([sys.executable, '-c', script], env=env) for _ in range(10)]
+        error = None
         while procs:
             for proc in procs:
                 retcode = proc.poll()
-                if retcode != None:
+                if retcode is not None:
+                    error = '{}\n{}'.format(proc.stderr, proc.stdout)
                     procs.remove(proc)
                     break
                 else:  # No process is done, wait a bit and check again.
@@ -367,7 +369,7 @@ class TestTrajectoryInfoCache(unittest.TestCase):
                 failed = True
                 break
 
-        self.assertTrue(not failed)
+        self.assertTrue(not failed, msg=error)
 
 if __name__ == "__main__":
     unittest.main()

--- a/pyemma/coordinates/tests/test_traj_info_cache.py
+++ b/pyemma/coordinates/tests/test_traj_info_cache.py
@@ -20,7 +20,7 @@ Created on 30.04.2015
 @author: marscher
 '''
 
-from __future__ import absolute_import
+from __future__ import absolute_import, print_function
 
 from tempfile import NamedTemporaryFile
 
@@ -347,18 +347,22 @@ class TestTrajectoryInfoCache(unittest.TestCase):
         import subprocess
         import sys
         import time
-        script = 'import pyemma; assert pyemma.config.cfg_dir=="{cfg_dir}", pyemma.config.cfg_dir; pyemma.coordinates.source({files})' \
+        script = 'import pyemma; pyemma.coordinates.source({files})' \
             .format(cfg_dir=self.work_dir, files=npy_files)
         failed = False
-        procs = [subprocess.Popen([sys.executable, '-c', script], env=env) for _ in range(10)]
+        procs = [subprocess.Popen([sys.executable, '-c', script], env=env,
+                                  stdout=subprocess.PIPE, stderr=subprocess.PIPE) for _ in range(10)]
         error = None
         while procs:
             for proc in procs:
                 retcode = proc.poll()
                 if retcode is not None:
-                    error = '{}\n{}'.format(proc.stderr, proc.stdout)
+                    if retcode != 0:
+                        stdout = proc.stdout.read()
+                        stderr = proc.stderr.read()
+                        error = '{};;{}'.format(stdout, stderr)
                     procs.remove(proc)
-                    break
+                    #break
                 else:  # No process is done, wait a bit and check again.
                     time.sleep(.1)
                     continue

--- a/pyemma/coordinates/tests/test_traj_info_cache.py
+++ b/pyemma/coordinates/tests/test_traj_info_cache.py
@@ -350,17 +350,17 @@ class TestTrajectoryInfoCache(unittest.TestCase):
         script = 'import pyemma; pyemma.coordinates.source({files})' \
             .format(cfg_dir=self.work_dir, files=npy_files)
         failed = False
-        procs = [subprocess.Popen([sys.executable, '-c', script], env=env,
-                                  stdout=subprocess.PIPE, stderr=subprocess.PIPE) for _ in range(10)]
+        procs = [subprocess.Popen([sys.executable, '-c', script], env=env) for _ in range(10)]
         error = None
         while procs:
             for proc in procs:
                 retcode = proc.poll()
                 if retcode is not None:
                     if retcode != 0:
-                        stdout = proc.stdout.read()
-                        stderr = proc.stderr.read()
-                        error = '{};;{}'.format(stdout, stderr)
+                        pass
+                        #stdout = proc.stdout.read()
+                        #stderr = proc.stderr.read()
+                        #error = '{};;{}'.format(stdout, stderr)
                     procs.remove(proc)
                     #break
                 else:  # No process is done, wait a bit and check again.

